src/components/AppProvider/AppProvider.tsx

```diff
@@ -45,89 +40,52 @@ export interface AppProviderProps extends AppBridgeOptions {
   UNSTABLE_telemetry?: TelemetryObject;
 }
 
-export class AppProvider extends React.Component<AppProviderProps, State> {
-  private stickyManager: StickyManager;
-  private scrollLockManager: ScrollLockManager;
-  private uniqueIdFactory: UniqueIdFactory;
-
-  constructor(props: AppProviderProps) {
-    super(props);
-    this.stickyManager = new StickyManager();
-    this.scrollLockManager = new ScrollLockManager();
-    this.uniqueIdFactory = new UniqueIdFactory(globalIdGeneratorFactory);
-
-    const {i18n, apiKey, shopOrigin, forceRedirect, linkComponent} = this.props;
-
-    // eslint-disable-next-line react/state-in-constructor
-    this.state = {
-      link: linkComponent,
-      intl: new I18n(i18n),
-      appBridge: createAppBridge({shopOrigin, apiKey, forceRedirect}),
-    };
-  }
-
-  componentDidMount() {
-    if (document != null) {
-      this.stickyManager.setContainer(document);
-    }
-  }
-
-  componentDidUpdate({
-    i18n: prevI18n,
-    linkComponent: prevLinkComponent,
-    apiKey: prevApiKey,
-    shopOrigin: prevShopOrigin,
-    forceRedirect: prevForceRedirect,
-  }: AppProviderProps) {
-    const {i18n, linkComponent, apiKey, shopOrigin, forceRedirect} = this.props;
-
-    if (
-      i18n === prevI18n &&
-      linkComponent === prevLinkComponent &&
-      apiKey === prevApiKey &&
-      shopOrigin === prevShopOrigin &&
-      forceRedirect === prevForceRedirect
-    ) {
-      return;
-    }
+export function AppProvider({
+  apiKey,
+  shopOrigin,
+  forceRedirect,
+  i18n: i18nProp,
+  linkComponent,
+  theme = {},
+  features = {},
+  UNSTABLE_telemetry,
+  children,
+}: AppProviderProps) {
+  const stickyManager = useLazyRef(() => new StickyManager());
+  const scrollLockManager = useLazyRef(() => new ScrollLockManager());
+  const uniqueIdFactory = useLazyRef(
+    () => new UniqueIdFactory(globalIdGeneratorFactory),
+  );
 
-    // eslint-disable-next-line react/no-did-update-set-state
-    this.setState({
-      link: linkComponent,
-      intl: new I18n(i18n),
-      appBridge: createAppBridge({shopOrigin, apiKey, forceRedirect}),
-    });
-  }
+  const appBridge = useMemo(
```

Comment:
https://reactjs.org/docs/hooks-reference.html#usememo suggests "You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render".

Will a bad thing happen if we change from using one instance of appBridge to a new one because react dropped the memoised value in the middle of the lifetime of an app? Perhaps sticking this in a ref and updating the ref if any of `[apiKey, shopOrigin, forceRedirect]` changes might work instead (this is a half-baked idea)

Same goes for the i18n instance, but I don't think there will be much of an issue with that using a new instance in the middle of the life of an app

Reply:
Nope nothing bad will happen from creating a new instance of appBridge IIRC, we're in the clear 👌